BeanMapper v2.4.0 fails if Spring Security dependency is lacking

Happens with v2.4.0, v.2.3.2 is safe.

```
java.lang.NoClassDefFoundError: Lorg/springframework/security/authentication/AuthenticationManager;

	at java.lang.Class.getDeclaredFields0(Native Method)
	at java.lang.Class.privateGetDeclaredFields(Class.java:2583)
	at java.lang.Class.getDeclaredFields(Class.java:1916)
	at org.springframework.util.ReflectionUtils.getDeclaredFields(ReflectionUtils.java:715)
	at org.springframework.util.ReflectionUtils.doWithFields(ReflectionUtils.java:687)
	at org.springframework.util.ReflectionUtils.doWithFields(ReflectionUtils.java:673)
	at org.springframework.boot.test.mock.mockito.DefinitionsParser.parse(DefinitionsParser.java:63)
	at org.springframework.boot.test.mock.mockito.MockitoPostProcessor.postProcessBeanFactory(MockitoPostProcessor.java:142)
	at org.springframework.boot.test.mock.mockito.MockitoPostProcessor.postProcessBeanFactory(MockitoPostProcessor.java:134)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:281)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:169)
	at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:687)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:525)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:693)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:360)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:303)
	at org.springframework.boot.test.context.SpringBootContextLoader.loadContext(SpringBootContextLoader.java:120)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:98)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:116)
	at org.springframework.test.context.support.DefaultTestContext.getApplicationContext(DefaultTestContext.java:83)
	at org.springframework.test.context.web.ServletTestExecutionListener.setUpRequestContextIfNecessary(ServletTestExecutionListener.java:189)
	at org.springframework.test.context.web.ServletTestExecutionListener.prepareTestInstance(ServletTestExecutionListener.java:131)
	at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:230)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.createTest(SpringJUnit4ClassRunner.java:228)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner$1.runReflectiveCall(SpringJUnit4ClassRunner.java:287)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.methodBlock(SpringJUnit4ClassRunner.java:289)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:247)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:94)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:191)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: java.lang.ClassNotFoundException: org.springframework.security.authentication.AuthenticationManager
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 51 more

```

--- a/pom.xml
+++ b/pom.xml
@@ -4,7 +4,7 @@
     
     <groupId>io.beanmapper</groupId>
     <artifactId>beanmapper-spring-boot-starter</artifactId>
-    <version>3.0.1</version>
+    <version>3.0.2-SNAPSHOT</version>
     <packaging>jar</packaging>
     <name>42 BeanMapper Spring Boot Starter</name>
     <description>Spring boot starter/autoconfig for BeanMapper</description>

--- a/src/main/java/io/beanmapper/autoconfigure/BeanMapperAutoConfig.java
+++ b/src/main/java/io/beanmapper/autoconfigure/BeanMapperAutoConfig.java
@@ -24,6 +24,7 @@ import io.beanmapper.spring.web.converter.StructuredJsonMessageConverter;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.beans.BeanInstantiationException;
+import org.springframework.beans.factory.NoSuchBeanDefinitionException;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.autoconfigure.AutoConfigureAfter;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
@@ -34,10 +35,9 @@ import org.springframework.boot.context.properties.EnableConfigurationProperties
 import org.springframework.context.ApplicationContext;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
-import org.springframework.context.annotation.Lazy;
 import org.springframework.data.repository.core.EntityInformation;
 import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter;
-import org.springframework.security.authentication.AuthenticationManager;
+import org.springframework.util.ClassUtils;
 import org.springframework.web.method.support.HandlerMethodArgumentResolver;
 import org.springframework.web.servlet.config.annotation.WebMvcConfigurerAdapter;
 
@@ -62,10 +62,6 @@ public class BeanMapperAutoConfig {
     private BeanMapperBuilderCustomizer builderCustomizer;
     @Autowired(required = false)
     private EntityManager entityManager;
-
-    @Lazy
-    @Autowired(required = false)
-    private AuthenticationManager authenticationManager;
 
     private ApplicationScanner collectionHandlerAppScanner;
     private ApplicationScanner beanConverterAppScanner;
@@ -97,14 +93,27 @@ public class BeanMapperAutoConfig {
         addCustomConverters(builder, packagePrefix);
         addCustomBeanPairs(builder);
         addAfterClearFlusher(builder);
-        setSecuredChecks(builder, packagePrefix);
+
+        if (ClassUtils.isPresent("org.springframework.security.authentication.AuthenticationManager", applicationContext.getClassLoader())) {
+            setSecuredChecks(builder, packagePrefix);
+        } else {
+            log.info("Spring Security is not present on the classpath. BeanMapper's @BeanLogicSecured and @BeanRoleSecured annotations will not be processed.");
+        }
+
         setUnproxy(builder);
         customize(builder);
         return builder.build();
     }
 
     private void setSecuredChecks(BeanMapperBuilder builder, String packagePrefix) {
-        if (authenticationManager == null || !props.getApplySecuredProperties()) {
+        try {
+            applicationContext.getBean(org.springframework.security.authentication.AuthenticationManager.class);
+        } catch (NoSuchBeanDefinitionException e) {
+            log.warn("No AuthenticationManager bean has been configured within your application. BeanMapper's @BeanLogicSecured and @BeanRoleSecured annotations can not be processed.");
+            return;
+        }
+        
+        if (!props.getApplySecuredProperties()) {
             return;
         }
         addLogicSecuredChecks(builder, packagePrefix);

--- a/src/test/java/io/beanmapper/autoconfigure/BeanMapperAutoConfigTest.java
+++ b/src/test/java/io/beanmapper/autoconfigure/BeanMapperAutoConfigTest.java
@@ -2,7 +2,9 @@ package io.beanmapper.autoconfigure;
 
 import static java.util.stream.Collectors.toList;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
 import static org.springframework.boot.test.util.EnvironmentTestUtils.addEnvironment;
 import static org.springframework.test.util.ReflectionTestUtils.getField;
 
@@ -15,6 +17,7 @@ import io.beanmapper.core.collections.CollectionHandler;
 import io.beanmapper.core.converter.BeanConverter;
 import io.beanmapper.core.unproxy.BeanUnproxy;
 import io.beanmapper.core.unproxy.DefaultBeanUnproxy;
+import io.beanmapper.spring.security.SpringRoleSecuredCheck;
 import io.beanmapper.spring.unproxy.HibernateAwareBeanUnproxy;
 import io.beanmapper.spring.web.MergedFormMethodArgumentResolver;
 
@@ -26,6 +29,7 @@ import org.springframework.boot.autoconfigure.web.servlet.WebMvcAutoConfiguratio
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 import org.springframework.mock.web.MockServletContext;
+import org.springframework.security.authentication.AuthenticationManager;
 import org.springframework.web.context.support.AnnotationConfigWebApplicationContext;
 import org.springframework.web.method.support.HandlerMethodArgumentResolver;
 import org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter;
@@ -84,6 +88,36 @@ public class BeanMapperAutoConfigTest {
         assertEquals(2, customCollectionHandlers.size());
         TestCollectionHandlerWithAppCtx collectionHandler = (TestCollectionHandlerWithAppCtx)mapper.getConfiguration().getCollectionHandlerFor(TestEntity2.class);
         assertNotNull(collectionHandler.getApplicationContext());
+    }
+
+    @Test
+    public void autoconfig_shouldSetSecurityChecks() {
+        loadApplicationContext(ConfigWithSecurity.class);
+        BeanMapper mapper = context.getBean(BeanMapper.class);
+
+        assertTrue(mapper.getConfiguration().getRoleSecuredCheck() instanceof SpringRoleSecuredCheck);
+    }
+
+
+
+    @Test
+    public void autoconfig_shouldNotSetSecurityChecks_ifSpringSecurityIsMissingFromClassPath() {
+        loadApplicationContext(ConfigWithSecurity.class, new NoSpringSecurityClassLoader());
+        BeanMapper mapper = context.getBean(BeanMapper.class);
+
+        assertFalse(mapper.getConfiguration().getRoleSecuredCheck() instanceof SpringRoleSecuredCheck);
+    }
+
+
+    @Configuration
+    static class ConfigWithSecurity {
+
+        @Bean
+        public AuthenticationManager authenticationManager() {
+            return authentication -> {
+                throw new UnsupportedOperationException("Stub!");
+            };
+        }
     }
 
     @Configuration
@@ -154,7 +188,16 @@ public class BeanMapperAutoConfigTest {
     }
 
     private void loadApplicationContext(Class<?> config, String... env) {
+        loadApplicationContext(config, null, env);
+    }
+
+    private void loadApplicationContext(Class<?> config, ClassLoader classLoader, String... env) {
         AnnotationConfigWebApplicationContext applicationContext = new AnnotationConfigWebApplicationContext();
+
+        if (classLoader != null) {
+            applicationContext.setClassLoader(classLoader);
+        }
+
         addEnvironment(applicationContext, env);
         addEnvironment(applicationContext, BEANMAPPER_PACKAGE_PREFIX_PROP);
         if (config != null) {

--- a/src/test/java/io/beanmapper/autoconfigure/NoSpringSecurityClassLoader.java
+++ b/src/test/java/io/beanmapper/autoconfigure/NoSpringSecurityClassLoader.java
@@ -1,0 +1,17 @@
+package io.beanmapper.autoconfigure;
+
+/**
+ * Mocked ClassLoader which does not load the class used to detect if Spring Security is on the classpath.
+ * This way, we can test if BeanMapperAutoConfig correctly omits the security-related features in such cases.
+ */
+public class NoSpringSecurityClassLoader extends ClassLoader {
+
+    @Override
+    public Class<?> loadClass(String name) throws ClassNotFoundException {
+        if (name != null && !name.equals("org.springframework.security.authentication.AuthenticationManager")) {
+            return super.loadClass(name);
+        }
+
+        throw new NoClassDefFoundError(name);
+    }
+}

